cloudcompchem/models.py

```diff
@@ -3,7 +3,111 @@
 from dataclasses import dataclass
 from functools import partial
 
-from .exceptions import DFTRequestValidationException
+from .exceptions import (
+    DFTRequestValidationException,
+    MoleculeSpinAndChargeViolationError,
+)
+
+ATOMIC_NUMBERS = {
+    "H": 1,
+    "He": 2,
+    "Li": 3,
+    "Be": 4,
+    "B": 5,
+    "C": 6,
+    "N": 7,
+    "O": 8,
+    "F": 9,
+    "Ne": 10,
+    "Na": 11,
+    "Mg": 12,
+    "Al": 13,
+    "Si": 14,
+    "P": 15,
+    "S": 16,
+    "Cl": 17,
+    "Ar": 18,
+    "K": 19,
+    "Ca": 20,
+    "Sc": 21,
+    "Ti": 22,
+    "V": 23,
+    "Cr": 24,
+    "Mn": 25,
+    "Fe": 26,
+    "Co": 27,
+    "Ni": 28,
+    "Cu": 29,
+    "Zn": 30,
+    "Ga": 31,
+    "Ge": 32,
+    "As": 33,
+    "Se": 34,
+    "Br": 35,
+    "Kr": 36,
+    "Rb": 37,
+    "Sr": 38,
+    "Y": 39,
+    "Zr": 40,
+    "Nb": 41,
+    "Mo": 42,
+    "Tc": 43,
+    "Ru": 44,
+    "Rh": 45,
+    "Pd": 46,
+    "Ag": 47,
+    "Cd": 48,
+    "In": 49,
+    "Sn": 50,
+    "Sb": 51,
+    "Te": 52,
+    "I": 53,
+    "Xe": 54,
+    "Cs": 55,
+    "Ba": 56,
+    "La": 57,
+    "Ce": 58,
+    "Pr": 59,
+    "Nd": 60,
+    "Pm": 61,
+    "Sm": 62,
+    "Eu": 63,
+    "Gd": 64,
+    "Tb": 65,
+    "Dy": 66,
+    "Ho": 67,
+    "Er": 68,
+    "Tm": 69,
+    "Yb": 70,
+    "Lu": 71,
+    "Hf": 72,
+    "Ta": 73,
+    "W": 74,
+    "Re": 75,
+    "Os": 76,
+    "Ir": 77,
+    "Pt": 78,
+    "Au": 79,
+    "Hg": 80,
+    "Tl": 81,
+    "Pb": 82,
+    "Bi": 83,
+    "Po": 84,
+    "At": 85,
+    "Rn": 86,
+    "Fr": 87,
+    "Ra": 88,
+    "Ac": 89,
+    "Th": 90,
+    "Pa": 91,
+    "U": 92,
+    "Np": 93,
+    "Pu": 94,
+    "Am": 95,
+    "Cm": 96,
+    "Bk": 97,
+    "Cf": 98,
+}
```

Comment:
This representation looks correct but we can benefit from applying a more restrictive type than `str` to encode our atoms in the type system. The rationale is that we want to early catch situations where we look up atomic numbers for an atom that doesn't exist, e.g `ATOMIC_NUMBER["Hello"]`. I've found `typing.Literal` to be quite useful in these scenarios.

In a similar vein, we want to also constrain our choice for determining the atomic number from our "list" of atomic symbols. That's because while it didn't seem like it affects your code, it is still _possible_ to make a mistake like this one:

```
    "U": 92,
    "Np": 92,  # woops, should have been 93
    "Pu": 94,
    "Am": 95,
```

A bug like this is hard to detect and could have serious consequences, so we might as well leverage a different data structure that makes it impossible for us to assign an invalid index to an element. A tuple, for example, let's us infer the atomic number just by looking up the index for a given symbol. Of course, this means you now have to ensure that your list is correct but that's also desirable from the dictionary keys anyway.

Putting it all together, here's my proposal:

```python
from typing import Literal, get_args

AtomSymbol: Literal["H", "He", ...]
ATOMIC_NUMBERS: dict[AtomSymbol, int] = {
    atom: position
    for position, atom in enumerate(get_args(AtomSymbol), start=1)
}

ATOMIC_NUMBERS["hello"]  # pyright is onto you!
ATOMIC_NUMBERS["He"]  # looks good, go ahead...
```

Not that we'd want to also update the type for the `symbol` field in the `Atom` dataclass which, again, let's us catch mistakes early on:

```diff
--- symbol: str
+++ symbol: AtomSymbol
```

<img width="330" alt="image" src="https://github.com/cloudlab-cmu/cloudcompchem/assets/111795938/6803f23b-c7b5-4b31-802e-8aef346138f7">

Let me know if this all makes sense!

Reply:
@yiwenwang47 I set up a PR targeting this one with this and a few other suggestions. Let me know what you think! https://github.com/cloudlab-cmu/cloudcompchem/pull/36